### PR TITLE
ruben/eng-306 hsm support

### PR DESCRIPTION
**Pull request by topether21**
HSM Instructions:

- Connect Mk3/Mk4 via USB
- Install  https://github.com/Coldcard/ckcc-protocol
- pip install 'ckcc-protocol[cli]' ( I needed to do sudo for this; permissions issue with hidapi used for usb comms)
- can now use via ckcc (or sudo ckcc)
- Edit your .env:
```
 USE_HSM=false
 HSM_WALLET_ADDRESS=bc1p.. (based on your derivation path, can use ckcc addr command to generate an address based on derivation path)
 HSM_CLI_PATH="ckcc" (or sudo ckcc)
 HSM_DERIVATION_PATH="m/86'/0'/0'/0/0"
```
- Run `npm run hsm:address [derivation-path]` to ensure the output matches your HSM_WALLET_ADDRESS
- Run` npm run hsm:policy` to generate a hsm_policy.json file
- Use ckcc to start hsm mode: ckcc hsm-start ./hsm_policy.json
- Confirm the policy and start HSM mode on your Coldcard screen
- Start hunting!

You can use `npm run hsm` to easy configuration.

image

### CODE REVIEW

**Comment by dannydeezy** [comment]
I ran `npm run hsm:policy` with the following values in my config:
```
GEMINI_API_KEY="<REDACTED>"
GEMINI_API_SECRET="<REDACTED>"
GEMINI_DEPOSIT_ADDRESS=
GEMINI_WITHDRAWAL_ADDRESS=<REDACTED>

OKX_API_KEY=<REDACTED>
OKX_API_SECRET=<REDACTED>
OKX_DEPOSIT_ADDRESS=
OKX_WITHDRAWAL_ADDRESS=<REDACTED>
OKX_API_PASSPHRASE=<REDACTED>!

USE_HSM=true
HSM_CLI_PATH="ckcc"
HSM_WALLET_ADDRESS=<REDACTED>
```

then policy generated had two rules: one for my `GEMINI_WITHDRAWAL_ADDRESS` and one for my `OKX_WITHDRAWAL_ADDRESS` - why would these be in the rules? we won't be sending there

File: scripts/generate_coldcard_hsm_policy.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        'COINBASE_EXCHANGE',
+        'COINBASE_PRIME',
+        'GEMINI',
+        'BFX',
+        'BINANCE',
+        'OKX',
+        'BYBIT',
+        'KUCOIN'
+    ].reduce((acc, curr) => {
+        const withdrawal_address_key = `${curr}_WITHDRAWAL_ADDRESS`;
```
Comment (dannydeezy) [inline scripts/generate_coldcard_hsm_policy.js:52]: I don't think withdrawal addresses should be on the whitelist - only the deposit addresses

**Comment by dannydeezy** [comment]
- [ ] Add instructions to go into coldcard itself in Advanced / Tools and Enable HSM

**Comment by dannydeezy** [comment]
why do we need to remove `name` from the `hsm_policy` items?

**Comment by dannydeezy** [comment]
Getting this error when trying to do `hsm-start`

```
ckcc hsm-start ./hsm_policy.json
561 bytes (start @ 0) to send from 'hsm_policy.json'
Uploading  [####################################]  100%


Coldcard Error: HSM Policy invalid: hsm.py:112: period: must be in range: [1..4320]
```

File: index.js [hunk 2]
```diff
@@ -329,11 +333,11 @@ async function run() {
         let decodeError = null;
         try {
             await decode_sign_and_send_psbt({
-                psbt: info.extraction_psbt,
+                psbt: 'cHNidP8BALACAAAAAw0LFe2nYZVhWGRFccsyZ1kWzineRsUq2xM/gJAqfHvYAAAAAAD/////T3KwcSU2a3qn8J9opvAs61NbbruFTmebo4ZCgIJ/sukAAAAAAP////939XUhgWBqk1icrzcEAFQmsv6hKPx9tlzuQi23gby5FwAAAAAA/////wFAwfwQAAAAACIAICKwk9O/uMVkMqVMZ2HyTe1d5B1YdAZxHnutVEG9LSRcAAAAAAABASIA4fUFAAAAABl2qRQLJTen1vPMZoyen6AwP/s8rW6bgYisIgYDK+NygB2EYN2lKuF4qtd0pUgAulb5Scl7ii9R4pkgnQYMDwVpQwAAAAAAAAAAAAEBIgDh9QUAAAAAGXapFIKLdGycxuL7uJOU9U/0Dv/SpLZRiKwiBgLZIpA7jrKVuyMvwDBNiM8XVaYBeRVvqPhJcPDUBy1fuwwPBWlDAAAAAAEAAAAAAQEiAOH1BQAAAAAZdqkUEYjZdEwP6Oj67HctDEgb+BQA+QmIrCIGAh+0I9QlEuSddanSdBUEenQ3gkPPeu4Hwbabp9oKA9QoDA8FaUMAAAAAAgAAAAABASIAICKwk9O/uMVkMqVMZ2HyTe1d5B1YdAZxHnutVEG9LSRcAA==',
```
Comment (dannydeezy) [inline index.js:336]: this needs to be removed right?

**Comment by topether21** [comment]
@dannydeezy https://coldcard.com/docs/hsm/rules/ I never tested it on a real device. I made the PRs before the cold card arrived. The name is not present in the rules. I will change the default period.

File: index.js (hunk 2 above)
Comment (topether21) [reply index.js:336]: yes

**Comment by dannydeezy** [comment]
running into this error when signing:
```
Coldcard Error: PSBT does not contain any key path information.
```
I believe we need to add `bip32Derivation` to the inputData of the psbt input

**Comment by dannydeezy** [comment]
> running into this error when signing:
> 
> ```
> Coldcard Error: PSBT does not contain any key path information.
> ```
> 
> I believe we need to add `bip32Derivation` to the inputData of the psbt input

i've fixed this!

**Comment by dannydeezy** [comment]
I'm going to commit some changes that make it work
First, add this to `.env`
```
# ckcc xpub
HSM_XPUB=xpub1
# ckcc xpub "m/44'/0'/0'/0/0"
HSM_CHILD_XPUB=xpub2
```

**Review by dannydeezy — APPROVED** [review]
This is working great, nice work!

**Comment by dannydeezy** [comment]
@topether21 to help them save on fees, we should direct them to use the segwit address so I suggest the following changes:

address generation command:
```
ckcc addr --segwit "m/84'/0'/0'/0/0"
```

use new path:
```
HSM_DERIVATION_PATH="m/84'/0'/0'/0/0"
```

I am testing this flow now to ensure it works

File: .env.sample [hunk 3]
```diff
@@ -3,6 +3,25 @@
...
+# Enable HSM support for enhanced security during the signing process.
+# When set to 'true', sat-hunter will utilize a connected HSM device (e.g., Coldcard) for transaction signing.
+# This ensures that private keys are managed within a secure, tamper-resistant hardware environment, offering an additional layer of security.
+# To activate HSM support, explicitly set this value to 'true'. Any other value (including leaving it unset) will default to using the software-based signing process or other wallet types as configured.
+# Important: If enabling HSM support, ensure the necessary HSM device setup and corresponding software (e.g., ckcc-protocol for Coldcard) are correctly installed and configured.
+# Use the 'HSM_WALLET_ADDRESS' field to specify the Bitcoin address associated with the HSM device.
+# HSM_CLI_PATH="ckcc" ## Optional
+# USE_HSM=false
+# HSM_WALLET_ADDRESS=bc1p...
+# HSM_DERIVATION_PATH="m/44'/0'/0'/0/0"
```
Comment (dannydeezy) [inline .env.sample:19]:
```suggestion
# HSM_DERIVATION_PATH="m/84'/0'/0'/0/0"
```

File: .env.sample [hunk 4]
```diff
@@ -3,6 +3,25 @@
...
+# Example: ckcc xpub
+# HSM_XPUB="xpub661..."
+# Example: ckcc xpub "m/44'/0'/0'/0/0"
```
Comment (dannydeezy) [inline .env.sample:22]:
```suggestion
# Example: ckcc xpub "m/84'/0'/0'/0/0"
```

File: .env.sample [hunk 5]
```diff
@@ -3,6 +3,25 @@
...
+
+# Enable HSM support for enhanced security during the signing process.
+# When set to 'true', sat-hunter will utilize a connected HSM device (e.g., Coldcard) for transaction signing.
+# This ensures that private keys are managed within a secure, tamper-resistant hardware environment, offering an additional layer of security.
+# To activate HSM support, explicitly set this value to 'true'. Any other value (including leaving it unset) will default to using the software-based signing process or other wallet types as configured.
+# Important: If enabling HSM support, ensure the necessary HSM device setup and corresponding software (e.g., ckcc-protocol for Coldcard) are correctly installed and configured.
+# Use the 'HSM_WALLET_ADDRESS' field to specify the Bitcoin address associated with the HSM device.
+# HSM_CLI_PATH="ckcc" ## Optional
+# USE_HSM=false
+# HSM_WALLET_ADDRESS=bc1p...
```
Comment (dannydeezy) [inline .env.sample:18]:
```suggestion
# Generate a segwit address with: ckcc addr --segwit "m/84'/0'/0'/0/0"
# HSM_WALLET_ADDRESS=bc1p...
```

File: scripts/get_coldcard_address.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+const { get_address_from_coldcard } = require('../hsm')
+
+require('dotenv').config({
+    override: true,
+})
+;(async () => {
+    console.log(await get_address_from_coldcard())
```
Comment (dannydeezy) [inline scripts/get_coldcard_address.js:7]: @topether21 need to pass in the derivation path here

Comment (topether21) [reply scripts/get_coldcard_address.js:7]: image
image

Comment (dannydeezy) [reply scripts/get_coldcard_address.js]: nice!

**Review by dannydeezy — APPROVED** [review]

**Review by habibitcoin — APPROVED** [review]
Amazing! Just going to add to the README.md before merging in

**Pull request merged**